### PR TITLE
document notifySplices and fix render; fixes #1439

### LINKED ISSUE #1439: Correct documentation of dom-repeat.render(); add docs for notifySplices
_From @wuyuanyi135 on November 19, 2015 4:47_

I was trying to mutate a items array of a dom-repeat with javascript and see if the template will update after the change. 

This is my test code, and I host it on c9 on Nov 19, 2015. Hope [this](https://myweb-wuyuanyi.c9users.io/filetable) helps.
```html
<!DOCTYPE html>

<link rel="import" href="../polymer/polymer.html" />

<dom-module id="file-list">
    <template>
        Your files:
        <table>
            <template id="table_template" is="dom-repeat" items="{{fileItems}}">
                <tr>
                    <td>{{index}}</td>
                    <td>{{item.fileName}}</td>
                    <td>{{item.mimeType}}</td>
                </tr>
            </template>

        </table>
        <input type="button" value="Add one!" on-click="buttonClicked" />
    </template>

    <script>
        Polymer({
            is: "file-list",

            properties: {
                fileItems: {
                    value: [{
                        type: Array,
                        fileName: "filename 1",
                        mimeType: "mime 1"
                    }, {
                        fileName: "filename 2",
                        mimeType: "mime 2"
                    }],
                    notify: true
                }
            },

            buttonClicked: function() {
                //this.fileItems.push({fileName:"filename x",mimeType:"mimeType x"});
                //this.$.table_template.render();
                this.push("fileItems",{fileName:"filename x",mimeType:"mimeType x"});
            }
        });
    </script>
</dom-module>
```

The [documentation page](https://www.polymer-project.org/1.0/docs/devguide/templates.html#dom-repeat) indicates that I need to use polymer function `this.push("fileItems",object)` rather than `fileItems.push(object)`, otherwise I need manually call `this.$.table_template.render()` to update the template.

What happened is, when I append an object to `fileItems` array, the `render()` throws an exception
```
Uncaught TypeError: Cannot read property '0' of undefined
Polymer.Collection._parseKey @ polymer.html:3900
Polymer.Collection.getItem @ polymer.html:3919
Polymer._stampInstance @ polymer.html:4328
Polymer._insertInstance @ polymer.html:4338
Polymer._applyFullRefresh @ polymer.html:4201
Polymer._render @ polymer.html:4134
Debouncer.complete @ polymer.html:1231
Polymer.Base.extend.flush @ polymer-mini.html:1152
Polymer.Templatizer._flushTemplates @ polymer.html:3634
Polymer.render @ polymer.html:4129
```

I roughly traced the problem, and find that in function `getKey` in line 3883 of `polymer.html`, `key = this.omap.get(item);` was executed. But `this.omap` only contains 2 elements (initial value). Therefore, the new appended elements cause `key` to be `undefined` and then lead to the exception. 

```javascript
getKey: function (item) {
    var key;
    if (item && typeof item == 'object') {
        key = this.omap.get(item);
    } else {
        key = this.pmap[item];
    }
    if (key != undefined) {
        return '#' + key;
    }
}
```

Did i use the proper way to invoke render()? How can I update `this.omap`?

_Copied from original issue: Polymer/polymer#3022_

### PR DESCRIPTION
**Pull request by kaycebasques**
@arthurevans WDYT?

### CODE REVIEW

File: 1.0/docs/devguide/templates.md [hunk 1]
```diff
@@ -277,6 +278,11 @@ different name for the index property.
     </template>
     {% endraw %}
 
+### Forcing synchronous renders {#synchronous-renders}
+
+Call [`render`](/{{versionPrefix}}/api/#dom-repeat:method-render) to force
+a `dom-repeat` template to synchronously render any changes to its data. 
+Normally changes are batched and rendered asynchronously.
```
Comment (arthurevans) [inline 1.0/docs/devguide/templates.md]: We should add the two use cases here:

https://github.com/Polymer/docs/issues/1439#issuecomment-158156925

Hmm... the first one isn't really a use case, it's more of a general description. I think you could replace it with these two:

- For unit testing, to ensure items have rendered before checking the generated DOM.
- To ensure a list of items have rendered before scrolling to a specific item.

And explicitly note that it does _not_ cause the dom-repeat to pick up changes to the array that were made without the Polymer array mutation methods. If you (or a third-party library) has mutated the array without using the Polymer array mutation methods, you need to use notifySplices (link to section). 

This is one of the #1 things that confuses people about using render.

File: 1.0/docs/devguide/properties.md [hunk 2]
```diff
@@ -613,6 +613,18 @@ Example:
       </script>
     </dom-module>
 
+#### Using native array mutation methods {#notifysplices}
+
+Whenever possible you should always use Polymer's 
+[array mutation methods](#array-mutation).
+However, in some scenarios you might be forced to mutate arrays using the native
+`Array.prototype` methods. For example, you may be using a third-party library
```
Comment (arthurevans) [inline 1.0/docs/devguide/properties.md]: This is a little wordy. What do you think of just:

"However, this isn't always possible." for the 2nd sentence here?

And maybe "that is not aware of" => "that does not use"?

File: 1.0/docs/devguide/templates.md [hunk 3]
```diff
@@ -150,6 +148,9 @@ function defined on the host element.
 By default, the `filter` and `sort` functions only run when the array itself
 is mutated (for example, by adding or removing items).
 
+If you change the `filter` or `sort` function and want to re-process the data,
```
Comment (arthurevans) [inline 1.0/docs/devguide/templates.md]: These lines are incorrect. If you _change_ the `filter` or `sort` function, they'll be re-run automatically. You can use `render` to re-run the `filter` or `sort` function when some unrelated piece of data changes (for example, if your element has a `sortOrder` property that changes how the `sort` function works, you can call `render` when `sortOrder` changes.

File: 1.0/docs/devguide/templates.md [hunk 4]
```diff
@@ -150,6 +148,9 @@ function defined on the host element.
 By default, the `filter` and `sort` functions only run when the array itself
 is mutated (for example, by adding or removing items).
```
Comment (arthurevans) [inline 1.0/docs/devguide/templates.md]: How about:

"when ..." =>
"when one of the following occurs:

*   The array itself is mutated (for example, by adding or removing items).
*   The `filter` or `sort` function is changed.

**Comment by arthurevans** [comment]
Done reviewing. Left a few comments.

File: 1.0/docs/devguide/properties.md (hunk 2 above)
Comment (kaycebasques) [inline 1.0/docs/devguide/properties.md]: :heavy_check_mark:

File: 1.0/docs/devguide/templates.md (hunk 3 above)
Comment (kaycebasques) [inline 1.0/docs/devguide/templates.md]: :heavy_check_mark:

File: 1.0/docs/devguide/templates.md (hunk 4 above)
Comment (kaycebasques) [inline 1.0/docs/devguide/templates.md]: :heavy_check_mark:

**Comment by kaycebasques** [comment]
OK @arthurevans WDYT?

File: 1.0/docs/devguide/templates.md [hunk 5]
```diff
@@ -277,6 +283,23 @@ different name for the index property.
...
+
+Call [`render`](/{% polymer_version_dir %}/api/#dom-repeat:method-render) 
+to force a `dom-repeat` template to synchronously render any changes to its 
+data. Normally changes are batched and rendered asynchronously. This is 
+useful in two scenarios:
+
+*   For unit testing, to ensure items have rendered before checking the 
+    generated DOM.
+*   To ensure a list of items have rendered before scrolling to a 
+    specific item.
```
Comment (arthurevans) [inline 1.0/docs/devguide/templates.md]: *   To rerun the `sort` or `filter` functions when an a piece of data changes _outside_ the array (for example, sort order or filter criteria).

File: 1.0/docs/devguide/templates.md [hunk 6]
```diff
@@ -277,6 +283,23 @@ different name for the index property.
     </template>
     {% endraw %}
 
+### Forcing synchronous renders {#synchronous-renders}
+
+Call [`render`](/{% polymer_version_dir %}/api/#dom-repeat:method-render) 
+to force a `dom-repeat` template to synchronously render any changes to its 
+data. Normally changes are batched and rendered asynchronously. This is
```
Comment (arthurevans) [inline 1.0/docs/devguide/templates.md]: Note that given the order of these sentences, "This is useful..." appears to refer to "Normally changes are batched..."

So probably, "This is useful in two scenarios:" => "Synchronous rendering has a performance cost, but can be useful in a few scenarios:"

**Comment by arthurevans** [comment]
Caught a couple of other things on 2nd pass. Left comments, otherwise LG.

File: 1.0/docs/devguide/templates.md (hunk 6 above)
Comment (kaycebasques) [inline 1.0/docs/devguide/templates.md]: :heavy_check_mark:

File: 1.0/docs/devguide/templates.md (hunk 5 above)
Comment (kaycebasques) [inline 1.0/docs/devguide/templates.md]: :heavy_check_mark:

**Pull request merged**